### PR TITLE
lib/vfscore: Avoid recursive locking in sys_rename

### PR DESCRIPTION
**Pull request by marcrittinghaus**
If the source and destination directories are the same in sys_rename, the vnode is locked two times. This is problematic with non-recursive mutexes. Also the double checks for access are not needed. This commit adds a check to make sure in this case, only a single lock is done and redundant checks are avoided.

### Prerequisite checklist

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.pl`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.pl) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.

### Base target

 - Architecture(s): [N/A]
 - Platform(s): [N/A]
 - Application(s): [N/A]

### Description of changes

If the source and destination directories are the same in sys_rename,
the vnode is locked two times. This is problematic with non-recursive
mutexes. Also the double checks for access are not needed. This commit
adds a check to make sure in this case, only a single lock is done and
redundant checks are avoided.

The issue can be encountered in `app-python3` when applying #476

### CODE REVIEW

File: lib/vfscore/syscalls.c [hunk 1]
```diff
@@ -839,20 +839,23 @@ sys_rename(char *src, char *dest)
...
-	vn_lock(dvp2);
+	if (dvp1 != dvp2) {
+		vn_lock(dvp2);
 
-	/* Source and destination directions should be writable) */
-	if ((error = vn_access(dvp1, VWRITE)) != 0)
-	    goto err3;
-	if ((error = vn_access(dvp2, VWRITE)) != 0)
-	    goto err3;
+		/* The source and dest must be same file system */
```
Comment (mariasfiraiala) [inline lib/vfscore/syscalls.c]:
```suggestion
		/* The source and destination must be in the same file system */
```

Not sure about the meaning of the previous comment, so I amended it a bit.

File: lib/vfscore/syscalls.c [hunk 2]
```diff
@@ -839,20 +839,23 @@ sys_rename(char *src, char *dest)
...
+		if (dvp1->v_mount != dvp2->v_mount) {
+			error = EXDEV;
+			goto err3;
+		}
 
-	/* The source and dest must be same file system */
-	if (dvp1->v_mount != dvp2->v_mount) {
-		error = EXDEV;
-		goto err3;
+		/* Source and destination directions should be writable) */
```
Comment (mariasfiraiala) [inline lib/vfscore/syscalls.c]:
```suggestion
		/* Source and destination directions should be writable */
```

Not sure about "directions" here. Is it ok?

**Review by mariasfiraiala — CHANGES_REQUESTED** [review]
Thank you @marcrittinghaus. I've tested this with `app-python3` and #476 and it worked alright. I left some (nit-picky) suggestions regarding the old comments. I'll add my tag, anyway, since, from my pov, the functionality is fine.

Reviewed-by: Maria Sfiraiala <maria.sfiraiala@gmail.com>

File: lib/vfscore/syscalls.c (hunk 2 above)
Comment (StefanJum) [reply lib/vfscore/syscalls.c]: I think it's `directories`.

**Review by StefanJum — APPROVED** [review]
Besides @mariasfiraiala 's comments everything looks fine to me, thanks.
I'll add the tag after @razvand takes a look.

**Comment by marcrittinghaus** [comment]
> Thank you @marcrittinghaus. I've tested this with `app-python3` and #476 and it worked alright. I left some (nit-picky) suggestions regarding the old comments. I'll add my tag, anyway, since, from my pov, the functionality is fine.
> 
> Reviewed-by: Maria Sfiraiala [maria.sfiraiala@gmail.com](mailto:maria.sfiraiala@gmail.com)

Thanks @mariasfiraiala for pointing this out. That was a good opportunity to fix this. I did not see this 🐮💩. It is fixed 👍🏻 

Thanks @StefanJum for testing 🙏🏻

**Review by razvand — COMMENTED** [review]
It's OK.

@StefanJum, please add your tag and then I'll do the final approval step.

**Review by StefanJum — APPROVED** [review]
All good, go @razvand.

Reviewed-by: Stefan Jumarea <stefanjumarea02@gmail.com>

**Review by razvand — APPROVED** [review]
Approved-by: Razvan Deaconescu <razvand@unikraft.io>

**Pull request closed without merging**